tests/unit/fair_queue_test.cc

```diff
@@ -101,7 +90,6 @@ class test_env {
     unsigned tick(unsigned n = 0) {
         unsigned processed = 0;
         while (true) {
-            _fg.replenish_capacity(_fg.replenished_ts() + std::chrono::microseconds(1));
             _fq.dispatch_requests();
```

Comment:
Do I understand correctly that after defining:
```
    virtual throttle::grab_result can_dispatch() const noexcept override {
        return throttle::grab_result::grabbed;
    }
```

in `request` class, then this line will dispatch all request that were queued before? When reading the code of `dispatch_requests()` I had the impression, that it is the case.


If so, are we leaving the code as it was (mainly `while(true)` part) just in case if anybody wanted to define another request type that do not allow to be dispatched immediately? Is this `while(true)` needed (of course if I did not misunderstood the code) ?

Reply:
I suspect, that this is the root cause of the failure of the test.

Let's take `test_fair_queue_longer_run_different_shares` as an example.

We firstly queue a lot of ops:
```
    for (int i = 0; i < 20000; ++i) {
        env.do_op(a, 1);
        env.do_op(b, 0);
    }
```

And then we want to let them in over time:

```
   for (int i = 0; i < 1000; ++i) {
        sleep(1ms).get();
        env.tick(3);
    }
```

However, because all requests can be dispatched (new implementation of `can_dispatch()`) and the first call to `tick(3)` invokes `dispatch_requests()` that submits all requests at once, then the second call to `tick(3)` is stuck in infinite loop.

When I added a debug print in the loop that calls `tick(3)`, then I saw:

```
23: Call for i = 0
23: Call for i = 1
<reactor stall errors>
```

Reply:
This test fails because there's no (easy) way to tell dispatch loop to stop dispatching. So I marked this PR as draft to fix it later